src/batch/src/executor2/hop_window.rs

```diff
@@ -0,0 +1,305 @@
+    schema: Schema,
+    time_col_idx: usize,
+    window_slide: IntervalUnit,
+    window_size: IntervalUnit,
+}
+
+impl BoxedExecutor2Builder for HopWindowExecutor2 {
+    fn new_boxed_executor2(source: &ExecutorBuilder) -> Result<BoxedExecutor2> {
+        ensure!(source.plan_node().get_children().len() == 1);
+        let hop_window_node = try_match_expand!(
+            source.plan_node().get_node_body().unwrap(),
+            NodeBody::HopWindow
+        )?;
+        let time_col = hop_window_node.get_time_col()?.column_idx as usize;
+        let window_slide = hop_window_node.get_window_slide()?.into();
+        let window_size = hop_window_node.get_window_size()?.into();
+        if let Some(child_plan) = source.plan_node.get_children().get(0) {
+            let child = source.clone_for_plan(child_plan).build2()?;
+            let schema = child
+                .schema()
+                .clone()
+                .into_fields()
+                .into_iter()
+                .chain([
+                    Field::with_name(DataType::Timestamp, "window_start"),
+                    Field::with_name(DataType::Timestamp, "window_end"),
+                ])
+                .collect();
+            return Ok(Box::new(HopWindowExecutor2::new(
+                child,
+                schema,
+                time_col,
+                window_slide,
+                window_size,
+                source.plan_node().get_identity().clone(),
+            )));
+        }
+        Err(InternalError("HopWindow must have one child".to_string()).into())
+    }
+}
+
+impl HopWindowExecutor2 {
+    fn new(
+        child: BoxedExecutor2,
+        schema: Schema,
+        time_col_idx: usize,
+        window_slide: IntervalUnit,
+        window_size: IntervalUnit,
+        identity: String,
+    ) -> Self {
+        Self {
+            child,
+            identity,
+            schema,
+            time_col_idx,
+            window_slide,
+            window_size,
+        }
+    }
+}
+
+impl Executor2 for HopWindowExecutor2 {
+    fn schema(&self) -> &Schema {
+        self.child.schema()
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl HopWindowExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(self: Box<Self>) {
+        let Self {
+            child,
+            time_col_idx,
+            window_slide,
+            window_size,
+            ..
+        } = *self;
+        let units = window_size
+            .exact_div(&window_slide)
+            .and_then(|x| NonZeroUsize::new(usize::try_from(x).ok()?))
+            .ok_or_else(|| {
+                RwError::from(ErrorCode::InternalError(format!(
+                    "window_size {} cannot be divided by window_slide {}",
+                    window_size, window_slide
+                )))
+            })?
+            .get();
```

Comment:
I think we should do this in frontend rather in executor?

Reply:
It’s hard to do every thing in the frontend. The result is useless in frontend.

Reply:
But the dividable check should be done in frontend rather than doing it on backend.

Reply:
This is not a check, we need the result.

Reply:
If you want, we can panic here :)

Reply:
Frontend don’t need the divided result, so it’s unavoidable that we should calculate it again.

Reply:
Is this check to verify following sql invalide:
```
SELECT window_start, window_end, count(), sum(tax), sum(quantity)
FROM HOP(lineitems, shipdate, INTERVAL '2' DAYS, INTERVAL '7' DAYS)
GROUP BY window_start;
```
because `2 days` can't be divided by `7 days`?